task-system-design.md

```diff
@@ -0,0 +1,268 @@
+  ctx.extenders,
+);
+
+ctx.setRuntimeEnvironment(env)
+```
+
+### 3.2. Task system
+
+```mermaid
+classDiagram
+  class TaskParam {
+    String name
+    String type
+    String description
+    T defaultValue
+    Boolean isOptional
+
+    validator(value): Boolean
+  }
+
+  class Task {
+    String name
+    String description
+    Record~string, TaskParam~ params
+    Boolean isSubTask
+
+    action(args, env, runSuper)
+  }
+
+  class OverrideTask {
+    Task parentTask
+  }
+
+  class RuntimeEnvironment {
+    - Extender extenders
+    + Record~string, Task~ tasks
+
+    + run(name, args)
+  }
+
+  Task --o TaskParam: has
+  OverrideTask --|> Task : extends
+  RuntimeEnvironment --o Task : has
+```
+
+### 3.2.1. Task
+In the task system, any logic can be a `Task`, which should have a name as the task identifier, a description , params defined, and a task handler. 
+
+In order to implement task override, an `OverrideTask` is needed, which has an additional property of `parentTask` compared to `Task`. When the `action` of `OverrideTask` is run, the `action` of `parentTask` will be passed as `runSuper`, so that the override of task can be implemented
+
+### 3.2.2. RuntimeEnvironment
+We now have a simple task definition that allows us to run some logic, but the goal of the task system is to have some simple tasks that combine to accomplish a complex task, so we need a struct that manages all the tasks
+
+`RuntimeEnvironment` is a module to help us manage the execution of `Task` and to share the context between `Task`. When a task is executed, this is passed into the action of task as `env` so that the context can be shared between tasks
+
+
+#### How to load tasks
+When `KuaiContext` calls `loadConfigAndTasks` to initialize the project, it will run the `kuai.config.ts` file. you can put the custom task in it, or refer to the custom task's file in `kuai.config.ts`, so that the task can be injected into `KuaiContext`. and these injected tasks will be passed as parameters when `KuaiContext` initializes `RuntimeEnvironment`.
+
+Some helper functions are also provided in `kuai/core` to help simplify the code
+
+**Possible usage**
+```ts
+// kuai/core/helper
+function task(name, description) {
+  const ctx = KuaiContext.getKuaiContext();
+  const loader = ctx.tasksLoader;
+  return loader.addTask(name, description)
+}
+
+// custom-task.ts
+
+import { task } from 'kuai/core'
+
+task(TASK_NAME)
+  .addParam("paramA", "a custom param", 'default value')
+  .action(async (args, env, runSuper) => {
+  // do something…
+  })
+
+// kuai.config.ts
+
+import 'custom-plugin.ts'
+```
+
+When a `Task` with the same name is registered multiple times, the subsequent `Task` is loaded as an `OverrideTask`, and the previous Task is saved as a `parentTask`
+
+```ts
+// custom-override.ts
+
+import { task } from 'kuai/core'
+
+task(TASK_NAME)
+  .action(async (args, env, runSuper) => {
+    // do something before parent task…
+    await runSuper()
```

Comment:
is the `runSuper` nullable?, If it is, tell the user only available when it is overrided.